File: main/src/ca/uwaterloo/flix/runtime/solver/api/predicate/FilterPredicate.scala
package ca.uwaterloo.flix.runtime.solver.api.predicate

import ca.uwaterloo.flix.runtime.solver.api.term.{Term, WildTerm}

/**
  * Represents a filter predicate with function `f` and arguments `terms`.
  */
class FilterPredicate(f: Array[AnyRef] => Boolean, args: Array[Term]) extends Predicate {

  // TODO: Maybe the args should just be variables?

  /**
    * Invariant: A filter predicate cannot have wild card terms as arguments.
    */
  for (t <- args) {
    if (t.isInstanceOf[WildTerm]) {
      throw new IllegalArgumentException("A filter predicate cannot take a wild card term as an argument.")
    }
  }

  /**
    * Returns a copy of this predicate.
    */
  override def copy(): Predicate = new FilterPredicate(f, args)

  /**
    * Returns the function.
    */
  def getFunction(): Array[AnyRef] => Boolean = f

  /**
    * Returns the arguments.
    */
  def getArguments(): Array[Term] = args

  /**
    * Returns a string representation of `this` predicate.
    */
  override def toString: String = "<<filter>>" + "(" + args.mkString(", ") + ")"

}

File: main/src/ca/uwaterloo/flix/runtime/solver/api/predicate/FilterPredicate.java
package ca.uwaterloo.flix.runtime.solver.api.predicate;

import ca.uwaterloo.flix.runtime.solver.api.ProxyObject;
import ca.uwaterloo.flix.runtime.solver.api.term.Term;
import flix.runtime.fixpoint.predicate.Predicate;

import java.util.function.Function;

/**
 * Represents a filter predicate with function `f` and arguments `terms`.
 */
public final class FilterPredicate implements Predicate {

    /**
     * The function.
     */
    private final Function<Object[], ProxyObject> function;

    /**
     * The function arguments.
     */
    private final Term[] arguments;

    /**
     * Constructs a filter predicate for the given function and function arguments.
     */
    public FilterPredicate(Function<Object[], ProxyObject> function, Term[] arguments) {
        if (function == null)
            throw new IllegalArgumentException("'function' must be non-null.");

        if (arguments == null)
            throw new IllegalArgumentException("'arguments' must be non-null.");

        this.function = function;
        this.arguments = arguments;
    }

    /**
     * Returns the function.
     */
    public Function<Object[], ProxyObject> getFunction() {
        return function;
    }

    /**
     * Returns the arguments.
     */
    public Term[] getArguments() {
        return arguments;
    }

    /**
     * Returns a human-readable representation of `this` predicate.
     */
    @Override
    public String toString() {
        return "if <filter>";
    }

}
